packages/notifier/test/test-publish-kit.js

```diff
@@ -1,275 +1,310 @@
+      { message: 'Cannot update state after termination.' },
+      `${methodName} fails after the final value`,
+    );
+  }
+};
+
+// TODO: Replace with test.macro once that works with prepare-test-env-ava.
+// https://github.com/avajs/ava/blob/main/docs/01-writing-tests.md#reusing-test-logic-through-macros
+// eslint-disable-next-line no-shadow
+const verifySubscribeAfter = async (t, makePublishKit) => {
   const { publisher, subscriber } = makePublishKit();
+  for (const badCount of [1n, 0, '', false, Symbol('symbol'), {}]) {
+    t.throws(
+      // @ts-ignore deliberate invalid arguments for testing
+      () => subscriber.subscribeAfter(badCount),
+      undefined,
+      `subscribeAfter must reject invalid publish count: ${typeof badCount} ${q(
+        badCount,
+      )}`,
+    );
+  }
+  const subFirstP = subscriber.subscribeAfter(-999n);
+  publisher.publish('published');
+  const subFirst = await subFirstP;
+  t.deepEqual(subFirst.head, { value: 'published', done: false });
+};
 
-  publisher.publish(undefined);
-  const subFirst = await subscriber.subscribeAfter();
+for (const [type, maker] of Object.entries(makers)) {
+  test(type, verifyPublishKit, maker);
+  test(`${type} subscribeAfter`, verifySubscribeAfter, maker);
+}
 
-  const pubFailure = Symbol('fail');
-  publisher.fail(pubFailure);
-  const subFinalAll = await Promise.all(
-    [subFirst.tail, subscriber.subscribeAfter(subFirst.publishCount)].map(
-      invertPromiseSettlement,
-    ),
-  );
-  const [subFinal] = subFinalAll;
+test('publish kit allows non-durable values', async t => {
+  const publishKit = makePublishKit();
+  const nonPassable = { [Symbol('key')]: Symbol('value'), method() {} };
+  await assertTransmission(t, publishKit, nonPassable);
+  await assertTransmission(t, publishKit, nonPassable, 'finish');
+  await assertTransmission(t, makePublishKit(), nonPassable, 'fail');
+});
+test('durable publish kit rejects non-durable values', async t => {
+  const makeDurablePublishKit = vivifyDurablePublishKit(
+    makeBaggage(),
+    'DurablePublishKit',
+  );
+  const publishKit = makeDurablePublishKit();
+  const { publisher } = publishKit;
+  const nonPassable = { [Symbol('key')]: Symbol('value') };
+  t.throws(() => publisher.publish(nonPassable));
+  t.throws(() => publisher.finish(nonPassable));
+  t.throws(() => publisher.fail(nonPassable));
+  await assertTransmission(t, publishKit, Symbol.for('value'));
+});
+
+test('durable publish kit upgrade trauma', async t => {
+  const baggage = makeBaggage();
+  const makeDurablePublishKit = vivifyDurablePublishKit(
+    baggage,
+    'DurablePublishKit',
+  );
+  const kit1 = makeDurablePublishKit();
+  const { publisher: pub1, subscriber: sub1 } = kit1;
+  const value = Symbol.for('value');
+  await assertTransmission(t, kit1, value);
+  // THEN A MIRACLE OCCURS...
+  // @ts-expect-error
+  // eslint-disable-next-line no-undef
+  const kit2 = recoverPublishKit(baggage);
```

Comment:
I don't know how to do this, and could really use help. It looks like https://github.com/Agoric/agoric-sdk/blob/94625d38c3bb5333b00a69dd3086b1ac13490f62/packages/zoe/test/swingsetTests/upgradeCoveredCall/test-coveredCall-service-upgrade.js pulls off something analogous, but it does so at the vat level, which seems much too far up the stack for this. [prepare-test-env-ava.js](https://github.com/Agoric/agoric-sdk/blob/b43e34bd41b45de0d4771f4ade276e4b05bbffd0/packages/SwingSet/tools/prepare-test-env-ava.js) exports a globalThis.VatData populated from `{ vom, cm, wpm } = makeFakeVirtualStuff(…)` in [prepare-test-env.js](https://github.com/Agoric/agoric-sdk/blob/b43e34bd41b45de0d4771f4ade276e4b05bbffd0/packages/SwingSet/tools/prepare-test-env.js), and I have a sense that I should be doing something with [fakeVirtualSupport.js](https://github.com/Agoric/agoric-sdk/blob/b43e34bd41b45de0d4771f4ade276e4b05bbffd0/packages/SwingSet/tools/fakeVirtualSupport.js) to replace vrm and/or vom and/or crm (getting objects with new identity from reloaded durable state)—and probably an an independent test file to avoid corruption from manipulating the VatData global—but I can't see how to do it.

Reply:
With `upgradeCoveredCall`, we wanted to upgrade the contract in a clean environment, so we could be sure there wasn't any leakage of object references from one instance to the next. 

I think for the notifier upgrade, what I'd expect to see is exercising the notifier in one vat, then allowing it to shut down. Then in a second vat, reviving the notifier, and seeing that it could still be used by its original owner, and  previously existing clients would be able to see the new updates.

The first `run()` call in `test-coveredCall-service-upgrade` does the first part by invoking the `buildV1` clause in `bootstrap-coveredCall-service-upgrade`, which starts the `v2BundleId` installation of coveredCall.  The second `run()` call invokes `upgradeV2`, which uses the adminNode from the first vat and explicitly upgrades it.

When the upgrade calls `vivifyDurablePublishKit` in a new vat, the effect is to redefine the behavior of the publishKit and associate it with objects stored in the baggage with the same kind identifier. I don't think its allowed to re-associate new behavior with an existing kind identifier, so this has to be done in a new vat with access to the existing baggage.